File: csharp/proverb/Proverb.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Proverb
{
    public static string[] Recite(string[] subjects)
    {
        var res = new List<string>();

        if (subjects.Length > 0) {
            for (int i = subjects.Length - 1; i >= 0; i--) {
                if (i > 0) {
                    res.Insert(0, $"For want of a {subjects[i - 1]} the {subjects[i]} was lost.");
                } else {
                    res.Add($"And all for the want of a {subjects[i]}.");
                }
            }
        }
        return res.ToArray();
    }
}

File: typescript/proverb/proverb.ts
const Proverb = (...args: string[]): string => {
  const res: string[] = []
  for (let i = 0; i < args.length - 1; i++) {
    res.push(`For want of a ${args[i]} the ${args[i + 1]} was lost.`)
  }
  return [...res, `And all for the want of a ${args[0]}.`].join('\n');
}

export default Proverb